### PR TITLE
Feature/orangefalcon3/connecting clients

### LINKED ISSUE #12: Get two clients in the same room
## Goals
- Implement the Connecting page and the Game page
  - Register them to our navigator, keeping in mind #9 
- Make the necessary connections to the server and verify that two clients can be placed into the same room

### PR DESCRIPTION
**Pull request by rohanputhukudy**
### Administrative Info
[GitHub Issue #12]

### Changes
What changes did you make?
- Created socket context to enable frontend to connect to the socket server
- Refactored the socket server to assign rooms _after_ the initial connection

### Testing
How did you confirm your changes worked? 
- Used two physical phones to verify both connection to the socket server and room assignment

### Confirmation of Change 
Upload a screenshot, if possible. Otherwise, please provide instructions on how to see the change.
- Replace ``http://localhost:3000`` in SocketContext.tsx  with ``http://your_IP_address:3000``
- Then run the server and the frontend

### CODE REVIEW

File: mobile/src/screens/HomeScreen.tsx [hunk 1]
```diff
@@ -1,17 +1,26 @@
...
 import { AppStylesheet } from "../styles/AppStylesheet";
 
 const HomeScreen = () => {
   const { userState } = useContext(UserContext);
+  const socket = useContext(SocketContext);
+  const navigation = useNavigation();
+
+  const connectToGame = () => {
+    socket.emit("assign to room");
+    navigation.navigate("Chess");
```
Comment (nicolaswl) [inline mobile/src/screens/HomeScreen.tsx]: Can we implement the "connecting page" for this PR?

File: mobile/src/contexts/SocketContext.tsx [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import React from "react";
+import io from "socket.io-client";
+
+export const socket = io("http://localhost:3000", {
```
Comment (nicolaswl) [inline mobile/src/contexts/SocketContext.tsx]: add this to the environment file

File: mobile/src/components/chess/Piece.tsx [hunk 3]
```diff
@@ -85,7 +85,6 @@ const Piece = ({ id, startPosition, chess, onTurn, enabled }: PieceProps) => {
       });
       if (move) {
         chess.move({ from, to });
-        onTurn();
```
Comment (nicolaswl) [inline mobile/src/components/chess/Piece.tsx]: Why did you remove this? Is it redundant because of the modification to the chess object?

File: backend/src/eventHandlers/loginHandler.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import type { LoginHandlerParams, RoomData } from "../types";
+
+module.exports = function ({ socket, io, roomsMap, boards }: LoginHandlerParams) {
+  socket.on("successful login", (username: string) => {
```
Comment (nicolaswl) [inline backend/src/eventHandlers/loginHandler.ts:4]: # Note
This is where we will take in the token for validation and decode it to get the username.

File: mobile/App.tsx [hunk 5]
```diff
@@ -3,16 +3,19 @@ import React from "react";
...
 
 const App = () => (
-  <UserProvider>
-    <AuthProvider>
-      <NavigationContainer>
-        <Navigator />
-      </NavigationContainer>
-    </AuthProvider>
-  </UserProvider>
+  <SocketContext.Provider value={socket}>
```
Comment (nicolaswl) [inline mobile/App.tsx]: Let's change this to follow the format of the other contexts with a Provider component

File: backend/src/types.ts [hunk 6]
```diff
@@ -11,12 +11,19 @@ type BoardState = {
   players: string[];
 };
 
+interface LoginHandlerParams {
```
Comment (nicolaswl) [inline backend/src/types.ts]: I'm not sure if this is the best way to layout the types. I think it might be better to create nested types rather than duplicating so many common parameters. I am going to ask Justin for more info about this!

**Review by nicolaswl — CHANGES_REQUESTED** [review]

File: backend/src/types.ts (hunk 6 above)
Comment (lynguyxn) [reply backend/src/types.ts]: we might be able to do something like
```
interface LoginHandlerParams {
  ...
}
interface HandlerParams extends LoginHandlerParams {
  username: string;
}
```

Comment (nicolaswl) [reply backend/src/types.ts]: Yeah this is what I was thinking too. Let's do that but just provide some better naming. Call LoginHandlerParams something more generic like BaseHandlerParams. Then HandlerParams could be named something like GameHandlerParams because it also has a username. Then for all our future unique handlers we can extend BaseHandlerParams.

File: mobile/src/screens/HomeScreen.tsx (hunk 1 above)
Comment (nicolaswl) [reply mobile/src/screens/HomeScreen.tsx]: We're going to do this in another PR

**Comment by orangefalcon3** [comment]
The issues in the comments have been fixed.

File: mobile/src/components/chess/Piece.tsx (hunk 3 above)
Comment (rohanputhukudy) [reply mobile/src/components/chess/Piece.tsx]: I had accidentally deleted this while removing the socket's move event logic; it should be back now

File: backend/src/types.ts [hunk 7]
```diff
@@ -11,12 +11,15 @@ type BoardState = {
...
   socket: Socket;
   io: Server;
-  username: string;
-  rooms: Map<string, RoomData>;
+  roomsMap: Map<string, RoomData>;
   boards: Map<string, BoardState>;
 }
 
-export { RoomData, BoardState, HandlerParams };
+interface HandlerParams extends BaseHandlerParams {
```
Comment (nicolaswl) [inline backend/src/types.ts]: Are we just using this for login? If so, we should name it something like LoginParams instead of HandlerParams. HandlerParms seems generic but the BaseHandlerParams is the generic.

Comment (rohanputhukudy) [reply backend/src/types.ts]: We use HandlerParams for everything _except_ login, since it just adds a username field to BaseHandlerParams. So it should be generic enough to apply to all the other handlers. It does still seem kinda generic but idk what other name could work.

**Review by nicolaswl — APPROVED** [review]

**Review by AidenHuang01 — APPROVED** [review]
Approve PR
